### PR TITLE
feat: streak recover mutation

### PR DESCRIPTION
**Pull request by ilasw**
Mutation checks for:
- User is logged;
- User lost a streak in the past 24h, using Redis cache;
- Current streak is <= 1
- User can afford the cost, find for prev recover using `UserStreakActions` query;

> Not sure if the last 3 updates are the best way to update the user.

### CODE REVIEW

File: src/schema/users.ts [hunk 1]
```diff
@@ -1866,6 +1878,77 @@ export const resolvers: IResolvers<unknown, BaseContext> = traceResolvers<
...
       };
     },
+    recoverStreak: async (
+      _,
+      __,
+      ctx: AuthContext,
+      info,
+    ): Promise<GQLUserStreak> => {
+      const { userId } = ctx;
+      const user = await ctx.con
```
Comment (rebelchris) [inline src/schema/users.ts:1888]: There's `@auth` directive on graphQL side we shouldn't need to manually verify again.

File: src/schema/users.ts [hunk 2]
```diff
@@ -1866,6 +1878,77 @@ export const resolvers: IResolvers<unknown, BaseContext> = traceResolvers<
...
+        StorageTopic.Streak,
+        StorageKey.Reset,
+        userId,
+      );
+      const oldStreakLength = Number(await getRedisObject(key));
+      const userDoesntHaveOldStreak =
+        !oldStreakLength || !isNumber(oldStreakLength);
+      if (userDoesntHaveOldStreak) {
+        throw new ValidationError('No streak to recover');
+      }
```
Comment (rebelchris) [inline src/schema/users.ts:1902]: Not 100% sure but isn't this 1-1 the same as for the query, maybe we can combine some logic in a fn?

File: src/schema/users.ts [hunk 3]
```diff
@@ -1866,6 +1878,77 @@ export const resolvers: IResolvers<unknown, BaseContext> = traceResolvers<
...
+
+      const lastUserRecoverAction = await ctx.con
+        .getRepository(UserStreakAction)
+        .findOneBy({
+          userId,
+          type: UserStreakActionType.Recover,
+        });
+      const isFirstRecover = !lastUserRecoverAction;
+      const recoverCost = isFirstRecover ? 0 : 25;
```
Comment (rebelchris) [inline src/schema/users.ts:1918]: Can we maybe use `reputationReasonAmount` here somehow?

Comment (rebelchris) [reply src/schema/users.ts:1918]: For the 25 I mean

**Review by rebelchris — CHANGES_REQUESTED** [review]
Minor feedback, but overall looks good.

File: src/schema/users.ts (hunk 1 above)
Comment (ilasw) [reply src/schema/users.ts:1888]: I still need the `user.reputation` object below, should I change `findOneByOrFail` with `findOneBy`?

Comment (rebelchris) [reply src/schema/users.ts:1888]: Oh right, in that case leave it yes :)

File: src/schema/users.ts [hunk 4]
```diff
@@ -1866,6 +1878,77 @@ export const resolvers: IResolvers<unknown, BaseContext> = traceResolvers<
...
+      });
+
+      await ctx.con.getRepository(UserStreak).update(
+        {
+          userId,
+        },
+        {
+          currentStreak: oldStreakLength + streak.current,
+        },
+      );
```
Comment (capJavert) [inline src/schema/users.ts:1948]: We can do this in transaction to have changes saved in atomic way in the db.

**Review by capJavert — COMMENTED** [review]
Good work, appreciate the tests. Had some optimization ideas of data/loading saving.

File: src/schema/users.ts [hunk 5]
```diff
@@ -1866,6 +1878,77 @@ export const resolvers: IResolvers<unknown, BaseContext> = traceResolvers<
...
+    recoverStreak: async (
+      _,
+      __,
+      ctx: AuthContext,
+      info,
+    ): Promise<GQLUserStreak> => {
+      const { userId } = ctx;
+      const user = await ctx.con
+        .getRepository(User)
+        .findOneByOrFail({ id: userId });
```
[the remark this answers is not in the document]
Comment (ilasw) [reply src/schema/users.ts:1890]: Sure! Updated ✅

File: src/schema/users.ts [hunk 6]
```diff
@@ -1866,6 +1878,77 @@ export const resolvers: IResolvers<unknown, BaseContext> = traceResolvers<
...
+        .findOneBy({
+          userId,
+          type: UserStreakActionType.Recover,
+        });
+      const isFirstRecover = !lastUserRecoverAction;
+      const recoverCost = isFirstRecover ? 0 : streakRecoverCost;
+      const userCanAfford = user.reputation >= recoverCost;
+
+      if (!userCanAfford) {
+        throw new ValidationError('Not enough reputation to recover streak');
```
[the remark this answers is not in the document]
Comment (ilasw) [reply src/schema/users.ts:1922]: Yes! ✅

File: src/schema/users.ts (hunk 4 above)
Comment (ilasw) [reply src/schema/users.ts:1942]: Updated ✅

**Review by capJavert — APPROVED** [review]

File: src/schema/users.ts (hunk 3 above)
Comment (ilasw) [reply src/schema/users.ts]: Updated! ✅

File: src/schema/users.ts (hunk 2 above)
Comment (ilasw) [reply src/schema/users.ts]: Updated! ✅

File: __tests__/users.ts [hunk 7]
```diff
@@ -583,6 +588,126 @@ describe('query userStreaks', () => {
...
+      await con.getRepository(UserStreak).update(
+        { userId: loggedUser },
+        {
+          currentStreak: 0,
+          lastViewAt: subDays(new Date(), 2),
+        },
+      );
+      await con
+        .getRepository(User)
+        .update({ id: loggedUser }, { reputation: 25 });
```
Comment (capJavert) [inline __tests__/users.ts:687]: Would we not use a const here instead of `25`?

Comment (capJavert) [reply __tests__/users.ts:687]: Just so tests go through if we change the points

File: src/schema/users.ts [hunk 8]
```diff
@@ -1866,6 +1880,71 @@ export const resolvers: IResolvers<unknown, BaseContext> = traceResolvers<
...
+      const [user, lastUserRecoverAction] = await Promise.all([
+        ctx.con.getRepository(User).findOneByOrFail({ id: userId }),
+        await ctx.con.getRepository(UserStreakAction).findOneBy({
+          userId,
+          type: UserStreakActionType.Recover,
+        }),
+      ]);
+      const isFirstRecover = !lastUserRecoverAction;
+      const recoverCost = isFirstRecover ? 0 : streakRecoverCost;
+      const userCanAfford = user.reputation >= recoverCost;
```
Comment (capJavert) [inline src/schema/users.ts:1912]: Can't we just use `reputationReasonAmount` here?

File: src/schema/users.ts [hunk 9]
```diff
@@ -1866,6 +1880,71 @@ export const resolvers: IResolvers<unknown, BaseContext> = traceResolvers<
...
+      }
+
+      const reputationEvent = {
+        grantToId: userId,
+        targetId: userId,
+        targetType: ReputationType.Streak,
+        reason: isFirstRecover
+          ? ReputationReason.StreakFirstRecovery
+          : ReputationReason.StreakRecover,
+        amount: recoverCost * -1,
```
Comment (capJavert) [inline src/schema/users.ts:1925]: Then also here we don't have to do `* -1`

File: src/schema/users.ts (hunk 8 above)
Comment (capJavert) [reply src/schema/users.ts:1912]: Its fine this way around if you prefer just need to replace `0` with `reputationReasonAmount`

**Review by rebelchris — APPROVED** [review]
Nothing blocking from my side.

**Review by nensidosari — APPROVED** [review]

**Pull request merged**